File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        res = []
        queue = deque([(root, 0)])
        prevNode = None
        while queue:
            node, depth = queue.popleft()
            if prevNode and depth > prevNode[1]:
                res.append(prevNode[0].val)

            prevNode = (node, depth)

            if node.left:
                queue.append((node.left, depth + 1))
            if node.right:
                queue.append((node.right, depth + 1))
                
        res.append(node.val)
        return res

File: Data Structures & Algorithms/check-completeness-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        deque<TreeNode*> queue = {root};
        int level = 0;
        bool last = false;
        bool early = false;
        while (!queue.empty()) {
            cout << queue.size() << endl;
            int length = queue.size();
            if (length != pow(2, level)) {
                last = true;
            }

            for (int x = 0; x < length; x++) {
                TreeNode* node = queue.front();
                queue.pop_front();
                if (node->left) {
                    if (early)
                        return false;
                    if (last)
                        return false;
                    
                    queue.push_back(node->left);
                } else {
                    cout << node->val << endl;
                    early = true;
                }

                if (node->right) {
                    if (early)
                        return false;
                    if (last)
                        return false;
                    queue.push_back(node->right);
                } else {
                    cout << node->val << endl;
                    early = true;
                }
            }
            level += 1;
        }
        return true;
    }
};